File: authenticate/src/main/java/com/example/authenticate/service/AuthService.java
package com.example.authenticate.service;

import com.example.authenticate.dto.AuthRequestDTO;
import com.example.authenticate.entity.User;
import com.example.authenticate.repository.UserRepository;
import com.example.authenticate.response.ApiResponse;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
public class AuthService {

    private final UserRepository repo;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;

    public AuthService(UserRepository repo, PasswordEncoder passwordEncoder, JwtService jwtService) {
        this.repo = repo;
        this.passwordEncoder = passwordEncoder;
        this.jwtService = jwtService;
    }

    public ApiResponse<String> register(AuthRequestDTO request) {
        if (repo.findByUsername(request.getUsername()).isPresent()) {
            return new ApiResponse<>(409, "Username already exists", null); // Conflict
        }

        User user = new User();
        user.setUsername(request.getUsername());
        user.setPassword(passwordEncoder.encode(request.getPassword())); // Encode password

        User saved = repo.save(user);
        String token = jwtService.generateToken(saved.getUsername());
        return new ApiResponse<>(201, "Registered successfully", token); // Created
    }

    public ApiResponse<String> login(AuthRequestDTO request) {
        return repo.findByUsername(request.getUsername())
                .filter(user -> passwordEncoder.matches(request.getPassword(), user.getPassword()))
                .map(user -> {
                    String token = jwtService.generateToken(user.getUsername());
                    return new ApiResponse<>(200, "Login successful", token); // OK
                })
                .orElse(new ApiResponse<>(401, "Invalid credentials", null)); // Unauthorized
    }
}

File: authenticate-service-new/src/auth/auth.service.ts
import { Injectable, ConflictException, UnauthorizedException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { JwtService } from '@nestjs/jwt';
import { ConfigService } from '@nestjs/config';
import { Repository } from 'typeorm';
import * as bcrypt from 'bcrypt';

import { User } from './entities/user.entity';
import { AuthRequestDto } from './dto/auth-request.dto';
import { ApiResponseDto } from '../common/dto/api-response.dto';

@Injectable()
export class AuthService {
  constructor(
    @InjectRepository(User)
    private readonly userRepository: Repository<User>,
    private readonly jwtService: JwtService,
    private readonly configService: ConfigService,
  ) {}

  async register(authRequest: AuthRequestDto): Promise<ApiResponseDto<string | null>> {
    const existingUser = await this.userRepository.findOne({
      where: { username: authRequest.username },
    });

    if (existingUser) {
      return new ApiResponseDto(409, 'Username already exists', null);
    }

    // Hash password
    const saltRounds = 10;
    const hashedPassword = await bcrypt.hash(authRequest.password, saltRounds);

    // Create and save user
    const user = this.userRepository.create({
      username: authRequest.username,
      password: hashedPassword,
    });

    const savedUser = await this.userRepository.save(user);

    // Generate JWT token
    const token = this.generateToken(savedUser.username, savedUser.id);

    return new ApiResponseDto(201, 'Registered successfully', token);
  }

  async login(authRequest: AuthRequestDto): Promise<ApiResponseDto<string | null>> {
    const user = await this.userRepository.findOne({
      where: { username: authRequest.username },
    });

    if (!user) {
      return new ApiResponseDto(401, 'Invalid credentials', null);
    }

    // Check password
    const isPasswordValid = await bcrypt.compare(authRequest.password, user.password);
    
    if (!isPasswordValid) {
      return new ApiResponseDto(401, 'Invalid credentials', null);
    }

    // Generate JWT token
    const token = this.generateToken(user.username, user.id);

    return new ApiResponseDto(200, 'Login successful', token);
  }

  private generateToken(username: string, userId: number): string {
    const payload = { username, sub: userId };
    return this.jwtService.sign(payload, {
      secret: this.configService.get<string>('jwt.secret'),
      expiresIn: this.configService.get<string>('jwt.expiresIn'),
    });
  }
} 